contrib/format-excel/src/main/java/org/apache/drill/exec/store/excel/ExcelBatchReader.java

```diff
@@ -0,0 +1,398 @@
+  @Override
+  public boolean open(FileSchemaNegotiator negotiator) {
+    verifyConfigOptions();
+    split = negotiator.split();
+    openFile(negotiator);
+    TupleMetadata schema = defineSchema();
+    this.loader = negotiator.build();
+    return true;
+  }
+
+  private void openFile(FileScanFramework.FileSchemaNegotiator negotiator) {
+    try {
+      String filePath = split.getPath().toString();
+      this.fsStream = negotiator.fileSystem().open(new Path(filePath));
+      this.workbook = new XSSFWorkbook(this.fsStream.getWrappedStream());
+    } catch (Exception e) {
+      throw UserException.dataReadError(e).message("Failed to open open input file: %s", split.getPath().toString()).message(e.getMessage()).build(logger);
+    }
+
+    // Evaluate formulae
+    if (readerConfig.evaluateFormulae) {
+      this.evaluator = workbook.getCreationHelper().createFormulaEvaluator();
+    }
+    this.workbook.setMissingCellPolicy(Row.MissingCellPolicy.CREATE_NULL_AS_BLANK);
+    this.sheet = getSheet();
+  }
+
+  /**
+   * This helper function gets the column headers.
+   * If the user specified no columns headers it populates an array with field names of field_n
+   */
+  private TupleMetadata defineSchema() {
+    // TODO Use case if there are no column headers
+    SchemaBuilder builder = new SchemaBuilder();
+    return getColumnHeaders(builder);
+  }
+
+  protected TupleMetadata getColumnHeaders(SchemaBuilder builder) {
+    //Get the field names
+    int columnCount = 0;
+    if (readerConfig.headerRow >= 0) {
+      columnCount = sheet.getRow(readerConfig.headerRow).getPhysicalNumberOfCells();
+    } else {
+      columnCount = sheet.getRow(0).getPhysicalNumberOfCells();
+    }
+    this.excelFieldNames = new ArrayList<>(columnCount);
+    this.rowIterator = sheet.iterator();
+
+    //If there are no headers, create columns names of field_n
+    if (readerConfig.headerRow == -1) {
+      String missingFieldName;
+      for (int i = 0; i < columnCount; i++) {
+        missingFieldName = MISSING_FIELD_NAME_HEADER + (i + 1);
+        ExcelFormatPlugin.makeColumn(builder, missingFieldName, TypeProtos.MinorType.VARCHAR);
+        excelFieldNames.add(i, missingFieldName);
+      }
+      return builder.buildSchema();
+    } else if (rowIterator.hasNext()) {
+      //Find the header row
+      while (this.lineCount < readerConfig.headerRow) {
+        Row row = rowIterator.next();
+        this.lineCount++;
+      }
+      //Get the header row and column count
+      Row row = rowIterator.next();
+      this.totalColumnCount = row.getLastCellNum();
+
+      //Read the header row
+      Iterator<Cell> cellIterator = row.cellIterator();
+      int colPosition = 0;
+      String tempColumnName = "";
+
+      while (cellIterator.hasNext()) {
+        Cell cell = cellIterator.next();
+        // TODO Potential NPE if cell evaluation is turned off
+        CellValue cellValue = evaluator.evaluate(cell);
+        switch (cellValue.getCellTypeEnum()) {
```

Comment:
Are `STRING` and `NUMERIC` the only two cases? If not, what do we do for the other cases?

Reply:
There are 2.5 data types:  `String`, `Numeric` .  The 0.5 data type are timestamps, which are treated as numeric.

Reply:
I now see the issue here... Basically, if Drill is attempting to read the headers, they will be one of the two options: numeric or string.  For a column header, this needs to be coerced into a String.  Excel treats the `timestamp` data type as a `numeric` so for the purpose of getting the column header name, it can be rolled up into the `numeric` data type.

Reply:
If the sheet has no column header, Drill assigns names of `field_n` for each column.  This is verified with this test:
https://github.com/apache/drill/blob/b9d502108eb8b5e199c8855c793ea34f7537aa1c/contrib/format-excel/src/test/java/org/apache/drill/exec/store/excel/TestExcelFormat.java#L234-252

Reply:
If the sheet has no column header, Drill assigns names of `field_n` for each column.  This is verified with this test:
https://github.com/apache/drill/blob/b9d502108eb8b5e199c8855c793ea34f7537aa1c/contrib/format-excel/src/test/java/org/apache/drill/exec/store/excel/TestExcelFormat.java#L234-L252